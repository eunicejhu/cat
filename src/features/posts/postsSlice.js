import { createSlice } from "@reduxjs/toolkit";
import initialState from "../../helpers/initialState";

const postsSlice = createSlice({
  name: "posts",
  initialState: initialState.posts,
  reducers: {},
});

export default postsSlice.reducer;
