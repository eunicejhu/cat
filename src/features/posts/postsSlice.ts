import { createSlice, nanoid } from "@reduxjs/toolkit";

export type State = Post[];
export interface Post {
  id: string;
  title: string;
  content: string;
}

export interface PostAddedAction {
  type: string;
  payload: Post;
}

export interface PostUpdatedAction {
  type: string;
  payload: Post;
}

const postsSlice = createSlice({
  name: "posts",
  initialState: [],
  reducers: {
    postAdded: {
      reducer(state: State, action: PostAddedAction) {
        state.push(action.payload);
      },
      prepare({ title, content }) {
        return { payload: { id: nanoid(), title, content } };
      },
    },
    postUpdated: (state: State, action: PostUpdatedAction) => {
      const { id, title, content } = action.payload;
      const existingPost = state.find((post) => post.id === id);
      if (existingPost) {
        existingPost.title = title;
        existingPost.content = content;
      }
    },
  },
});
export const { postAdded, postUpdated } = postsSlice.actions;
export default postsSlice.reducer;
