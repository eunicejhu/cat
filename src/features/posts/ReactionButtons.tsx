import React from "react";
import { useDispatch } from "react-redux";
import { Post, reactionAdded } from "./postsSlice";

const reactionEmoji = {
  thumbsUp: "👍",
  hooray: "🎉",
  heart: "❤️",
  rocket: "🚀",
  eyes: "👀",
};

type Props = { post: Post };
type Name = keyof Props["post"]["reactions"];
const ReactionButtons: React.FC<Props> = ({
  post: { reactions, id },
}: Props) => {
  const dispatch = useDispatch();
  const reactionButtons = Object.entries(reactionEmoji).map(([name, emoji]) => (
    <button
      key={name}
      type="button"
      name={name}
      onClick={() =>
        dispatch(reactionAdded({ postId: id, reaction: name as Name }))
      }
    >
      {emoji} {reactions[name as Name]}
    </button>
  ));
  return <div className="reactions-container">{reactionButtons}</div>;
};

export default ReactionButtons;
