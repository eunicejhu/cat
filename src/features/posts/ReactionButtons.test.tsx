import React from "react";
import { fireEvent } from "@testing-library/react";
import { useDispatch } from "react-redux";
import ReactionButtons from "./ReactionButtons";
import { INITIAL_STATE } from "../../test/mock_data";
import renderWithStoreAndRouter from "../../test/renderWithStoreAndRouter";
import { reactionAdded } from "./postsSlice";

jest.mock("react-redux", () => ({
  ...(jest.requireActual("react-redux") as {}),
  useDispatch: jest.fn(),
}));
jest.mock("./postsSlice.ts", () => ({
  ...(jest.requireActual("./postsSlice.ts") as {}),
  reactionAdded: jest.fn(),
}));

test("render correctly", () => {
  const { asFragment } = renderWithStoreAndRouter(
    <ReactionButtons post={INITIAL_STATE.posts[0]} />
  );
  expect(asFragment()).toMatchSnapshot();
});

test("click reaction", async () => {
  const dispatch = jest.fn();
  (useDispatch as jest.Mock).mockReturnValue(dispatch);
  const { container } = renderWithStoreAndRouter(
    <ReactionButtons post={INITIAL_STATE.posts[0]} />
  );
  fireEvent.click(container.querySelector("[name=heart]") as Element);
  expect(dispatch).toHaveBeenCalled();
  expect(reactionAdded).toHaveBeenCalled();
});
