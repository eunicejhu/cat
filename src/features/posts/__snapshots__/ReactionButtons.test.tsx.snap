// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactionButtons test render correctly 1`] = `
<DocumentFragment>
  <div
    class="reactions-container"
  >
    <button
      name="thumbsUp"
      type="button"
    >
      👍 0
    </button>
    <button
      name="hooray"
      type="button"
    >
      🎉 0
    </button>
    <button
      name="heart"
      type="button"
    >
      ❤️ 4
    </button>
    <button
      name="rocket"
      type="button"
    >
      🚀 0
    </button>
    <button
      name="eyes"
      type="button"
    >
      👀 0
    </button>
  </div>
</DocumentFragment>
`;
